File: algorithms/java/Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int curDig;
        boolean carry = false;
        ListNode result, curNode;

        curDig = l1.val + l2.val;
        if (curDig > 9) {
            carry = true;
            curDig = curDig % 10;
        }
        result = new ListNode(curDig);
        curNode = result;

        while (l1.next != null || l2.next != null) {
            curDig = 0;
            if (carry) {
                curDig = 1;
            }
            if (l1.next != null) {
                l1 = l1.next;
                curDig += l1.val;
            }
            if (l2.next != null) {
                l2 = l2.next;
                curDig += l2.val;
            }
            if (curDig > 9) {
                carry = true;
                curDig = curDig % 10;
            } else {
                carry = false;
            }
            curNode.next = new ListNode(curDig);
            curNode = curNode.next;
        }
        if (carry) {
            curNode.next = new ListNode(1);
            curNode = curNode.next;
        }
        return result;
    }
}

File: algorithms/python/Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        digit = 1
        total = l1.val + l2.val
        while (l1.next):
            l1 = l1.next
            total += l1.val * (10 ** digit)
            digit += 1
        digit = 1
        while (l2.next):
            l2 = l2.next
            total += l2.val * (10 ** digit)
            digit += 1
        curNode = ListNode(0, None)
        ret = curNode
        while (total > 0):
            curNode.val = total % 10
            total = total // 10
            if (total > 0):
                curNode.next = ListNode(0, None)
                curNode = curNode.next
        return ret